osquery/core/init.cpp

```diff
@@ -714,19 +612,60 @@ void Initializer::start() const {
   EventFactory::delay();
 }
 
-void Initializer::waitThenShutdown() {
-  {
-    RecursiveLock lock(shutdown_mutex_);
-    if (shutdown_ != nullptr) {
-      // Copy the callable, then remove it, prevent callable recursion.
-      auto shutdown = shutdown_;
-      shutdown_ = nullptr;
-
-      // Call the shutdown callable.
-      shutdown();
+/**
+ * This is a small interruptable thread implementation.
+ *
+ * The goal is to wait until interrupted or an alarm timeout. If the timeout
+ * occures then osquery is stuck shutting and and we force-terminate.
+ */
+class AlarmRunnable : public InterruptableRunnable {
+ public:
+  /// Thread entry point.
+  void run() {
+    size_t waited = 0;
+    while (true) {
+      if (interrupted()) {
+        return;
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(200));
+      waited += 200;
+      if (waited > FLAGS_alarm_timeout * 1000) {
+        Initializer::shutdownNow(EXIT_CATASTROPHIC);
+      }
     }
   }
 
+ private:
+  /// No custom stop logic.
+  void stop() {}
+};
+
+/// Graceful shutdown request conditional var.
+std::condition_variable kShutdownRequestCV;
+
+void Initializer::waitForShutdown() const {
+  std::mutex shutdown_request_mutex;
+
+  // Attempt to be the only place in code where a join is attempted.
+  std::unique_lock<std::mutex> lock(shutdown_request_mutex);
+  kShutdownRequestCV.wait(lock);
+}
+
+int Initializer::shutdown(int retcode) const {
+  // Should only be called from main thread.
+  auto current_thread_id = std::this_thread::get_id();
+  if (current_thread_id != kMainThreadId) {
+    // Unintended usage.
+    throw std::runtime_error("Requested shutdown from service thread");
+  }
+
+  // Create an alarm thread, which will force-stop the process.
+  AlarmRunnable alarm_runnable;
+  auto alarm_thread = std::make_unique<std::thread>(
+      std::bind(&AlarmRunnable::run, &alarm_runnable));
+
+  // Request that all services stop.
+  Dispatcher::stopServices();
   // Attempt to be the only place in code where a join is attempted.
```

Comment:
Largely a nit: Hrm, I could be misunderstanding, but you have the same comment on line 649 as is here, but it doesn't look like `waitForShutdown` is `join`ing anything, rather just waiting for the mutex. Is that still considered a `join`? Would it make sense to change the comments?

edit: I think I'm understanding now, line 649 is the join for the watcher, where as this join is for the dispatched threads. Would it be worth changing the comments to specify _what_ is being joined?